File: prodcon_ipc/abstract_ipc.py
# -*- coding: utf-8 -*-

# Copyright (C) 2020 Adrian Böckenkamp
# This code is licensed under the BSD 3-Clause license (see LICENSE for details).

import logzero
import PyQt5.QtCore


class AbstractIPC(object):
    """
    Encapsulates code that both the producer and the consumer requires.
    """
    def __init__(self, id, key_file_path=None, log=True):
        """
        Creates the underlying system resources.

        :param id: Unique system-wide unique name (str) of shared memory; this name is also used to create the unique
        names for the two system-semapores `$id + "_sem_full"` and `$id + "_sem_empty"`
        :param key_file_path: Optional file path to a file typically named "shared_memory.key" whose first line is used
        as the unique ID/name of the shared memory IF this file exists, can be empty (the default) which then uses `id`
        (first parameter)
        :param log: `True` to enable logging using `logzero.logger`, `False` otherwise
        """
        self._log = log
        self._transaction_started = False
        if self._log:
            logzero.logger.info("Using PyQt v" + PyQt5.QtCore.PYQT_VERSION_STR + " and Qt v" +
                                PyQt5.QtCore.QT_VERSION_STR)

        self._file_key = self._load_key(key_file_path)
        self._shared_memory = PyQt5.QtCore.QSharedMemory(self._file_key if self._file_key else str(id))
        if self._log:
            logzero.logger.debug("Creating shared memory with key=\"" + self._shared_memory.key() + "\" (" +
                                 ("loaded from file)" if self._file_key else "hardcoded)"))
        self._sem_empty = PyQt5.QtCore.QSystemSemaphore(str(id) + "_sem_empty", 1, PyQt5.QtCore.QSystemSemaphore.Create)
        self._sem_full = PyQt5.QtCore.QSystemSemaphore(str(id) + "_sem_full", 0, PyQt5.QtCore.QSystemSemaphore.Create)

    def _load_key(self, path):
        """
        Alternatively loads the shared memory's name from a file named `SHARED_MEMORY_KEY_FILE`.

        :param path: Path to file whose first name contains the unique name; the rest is ignored
        :return: Unique name of shared memory or None if such a file did not exist
        """
        if path is None:
            return None
        # noinspection PyBroadException
        try:
            with open(path) as fp:
                line = fp.readline().strip()
                if fp.readline() and self._log:
                    logzero.logger.warn("Ignoring residual lines in " + path)
                return line
        except:
            pass
        return None

File: src/prodcon_ipc/abstract_ipc.cpp
// Copyright (C) 2020 Adrian Böckenkamp
// This code is licensed under the BSD 3-Clause license (see LICENSE for details).

#ifndef ABSTRACT_IPC_CPP
#define ABSTRACT_IPC_CPP

#include "abstract_ipc.h"

#include <QFile>
#include <QTextStream>
#include <QDebug>

AbstractIPC::AbstractIPC(const QString &id, const QString &key_file_path, bool log_debug) : log(log_debug),
  shared_memory(id),
  sem_empty(id + "_sem_empty", 1, QSystemSemaphore::AccessMode::Create),
  sem_full(id + "_sem_full", 0, QSystemSemaphore::AccessMode::Create)
{
  // Set unique name (key) of shared memory handle, use file if exiting and otherwise hardcoded
  // name:
  file_key = loadKey(key_file_path);
  if (!file_key.isEmpty()) {
    shared_memory.setKey(file_key);
  }
}

AbstractIPC::~AbstractIPC() { }

QString AbstractIPC::loadKey(const QString path) const
{
  if (path.isEmpty()) {
    return QString();
  }
  QFile file(path);
  if (file.open(QIODevice::ReadOnly)) {
    QTextStream in_str(&file);
    if (!in_str.atEnd()) {
      QString line = in_str.readLine().trimmed();
      in_str.readLine();
      if (!in_str.atEnd() && log) {
        qDebug() << "Ignoring residual lines in " << path;
      }
      return line;
    }
    file.close();
  }
  return QString();
}

void AbstractIPC::detach()
{
  // See also: https://doc.qt.io/qt-5/qsharedmemory.html#details
  if (!shared_memory.detach() && log) {
    qDebug() << "Unable to detach from shared memory.";
  }
}

#endif // ABSTRACT_IPC_CPP
